runtime/job-workers/client-functions.cpp

```diff
@@ -2,26 +2,92 @@
 // Copyright (c) 2021 LLC «V Kontakte»
 // Distributed under the GPL v3 License, see LICENSE.notice.txt
 
+#include "common/algorithms/clamp.h"
+
+#include "runtime/instance-copy-processor.h"
+#include "runtime/job-workers/job-interface.h"
 #include "runtime/job-workers/job-message.h"
 #include "runtime/job-workers/processing-jobs.h"
 #include "runtime/job-workers/shared-memory-manager.h"
-#include "runtime/instance-copy-processor.h"
-#include "runtime/net_events.h"
+#include "runtime/resumable.h"
 
 #include "server/job-workers/job-worker-client.h"
 #include "server/job-workers/job-stats.h"
+#include "server/php-engine-vars.h"
 
 #include "runtime/job-workers/client-functions.h"
 
-int64_t f$kphp_job_worker_start(const class_instance<C$KphpJobWorkerRequest> &request) noexcept {
-  if (request.is_null()) {
-    php_warning("Can't send job: the request shouldn't be null");
-    return -1;
+class job_resumable : public Resumable {
+public:
+  using ReturnT = class_instance<C$KphpJobWorkerResponse>;
+
+  explicit job_resumable(int job_id)
+    : job_id(job_id) {}
+
+protected:
+  bool run() final {
+    const class_instance<C$KphpJobWorkerResponse> &res = vk::singleton<job_workers::ProcessingJobs>::get().withdraw(job_id);
+    RETURN(res);
+  }
+private:
+  int job_id;
+};
+
+class kphp_job_worker_wait_resumable : public Resumable {
+public:
+  using ReturnT = class_instance<C$KphpJobWorkerResponse>;
+
+  kphp_job_worker_wait_resumable(int64_t resumable_id, double timeout)
+    : resumable_id(resumable_id)
+    , timeout(timeout) {}
+
+protected:
+  bool run() final {
+    bool ready{false};
+
+    RESUMABLE_BEGIN
+      ready = wait_without_result(resumable_id, timeout);
+      TRY_WAIT(kphp_job_worker_wait_resumable_label_0, ready, bool);
+      if (!ready) {
+        RETURN({});
+      }
+
+      Storage *input = get_forked_storage(resumable_id);
+      if (input->tag == 0) {
+        php_warning("Job result already was gotten\n");
+        RETURN({});
+      }
+      if (input->tag != Storage::tagger<ReturnT>::get_tag()) {
+        php_warning("Not a job request\n");
+        RETURN({});
+      }
+
+      const ReturnT &res = input->load<ReturnT>();
+      php_assert (CurException.is_null());
+
+      RETURN(res);
+    RESUMABLE_END
   }
+private:
+  int64_t resumable_id;
+  double timeout;
+};
+
+int64_t f$kphp_job_worker_start(const class_instance<C$KphpJobWorkerRequest> &request, double timeout) noexcept {
   if (!f$is_kphp_job_workers_enabled()) {
     php_warning("Can't send job: job workers disabled");
     return -1;
   }
+  if (request.is_null()) {
+    php_warning("Can't send job: the request shouldn't be null");
+    return -1;
+  }
+  if (timeout < 0) {
+    timeout = DEFAULT_SCRIPT_TIMEOUT;
+  } else if (timeout < 1 || timeout > MAX_SCRIPT_TIMEOUT) {
```

Comment:
I guess 1 second is too much for minimum.. 
May be 1 ms is more reasonable?

Reply:
If script execution timeout < 0.2 sec +- eps, script will be terminated before the start:

https://github.com/VKCOM/kphp/blob/80d895840cfd88194964f182aa8485d178974c78/server/php-worker.cpp#L161